sbat.c

```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
+				   badly parsed component_generation will be treated as zero 
+				*/
+				UINT16 sbat_gen = atoi(entry->component_generation);
+				UINT16 sbat_var_gen = atoi(sbat_var_entry->component_generation);
+				if ( sbat_gen < sbat_var_gen ) {
+					dprint(L"component's %a generation: %d. Conflicts with SBAT variable generation %d\n",
+					       entry->component_name, sbat_gen, sbat_var_gen );
+					LogError(L"image did not pass SBAT verification\n");
+					return EFI_SECURITY_VIOLATION;
+				}
+			}
+		}
+	}
+	dprint(L"all entries from SBAT section verified\n");
+	return EFI_SUCCESS;
+}
+
+static BOOLEAN
+is_utf8_bom(CHAR8 *buf, size_t bufsize)
+{
+	unsigned char bom[] = { 0xEF, 0xBB, 0xBF };
+	return !CompareMem(buf, bom, MIN(sizeof(bom), bufsize));
+}
+
+static struct sbat_var *
+new_entry(const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new_entry = AllocatePool(sizeof(*new_entry));
+
+	INIT_LIST_HEAD(&new_entry->list); 
+	new_entry->component_generation = comp_gen;
+	new_entry->component_name_size = comp_name_size;
+	new_entry->component_name = comp_name;
+	return new_entry;
+}
+
+static int
+add_entry(list_t *list, const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new;
+	new = new_entry(comp_gen, comp_name_size, comp_name);
+	if (!new)
+		return -1;
+
+	list_add_tail(&new->list, list);
+	return 0;
+}
+
+static void
+clean_up_vars(list_t *entries)
+{
+	list_t *pos = NULL, *tmp = NULL;
+	for_each_sbat_var_safe(pos, tmp, entries) {
+		struct sbat_var *entry;
+
+		entry = list_entry(pos, struct sbat_var, list);
+		list_del(&entry->list);
+		if (entry->component_generation)
+			FreePool((CHAR8 *)entry->component_generation);
+		if (entry->component_name_size)
+			FreePool((CHAR8 *)entry->component_name_size);
+		if (entry->component_name)
+			FreePool((CHAR8 *)entry->component_name);
+		FreePool(entry);
+	}
+}
+
+int
+parse_sbat_var(list_t *entries)
+{
+	UINT8 *data = 0;
+	UINTN datasize, i;
+	EFI_STATUS efi_status;
+	int rc;
+
+	INIT_LIST_HEAD(entries); 
+	efi_status = get_variable(L"SBAT", &data, &datasize, SHIM_LOCK_GUID);
+	if (EFI_ERROR(efi_status)) {
+		return -1;
+	}
+
+	CHAR8 *start = (CHAR8 *)data;
+	CHAR8 *end = (CHAR8 *)data + datasize;
+	if (is_utf8_bom(start, datasize))
+		start += 3;
+	while ((*end == '\r' || *end == '\n') && end < start)
+		end--;
+	*end = '\0';
+
+	dprint(L"SBAT variable data:\n");
+
+	rc = 0;
+	char delim;
+	while (start) {
+		const CHAR8 *fields[3] = { NULL, };
+		for (i = 0; i < 4; i++) {
+			const CHAR8 *tmp;
+			/* on fourth iteration we check if we had extra stuff on line while parsing last meaningful extraction of data
+			   if delimeter on 3rd iteration was ',', this means, we have comments after component_name, so let's drop them
+			*/
+			if (i == 3 && start[0] != '\0') {
+				if (delim == ',') {
+					start = get_sbat_field(start, end, &tmp, '\n');
+					break;
+				}
+				else
```

Comment:
Why do you need the else clause if you are breaking in both cases ?

Reply:
Also, if we are expecting 3 or 4 fields anyways I'm not sure that implementing it as a for loop is more readable instead of just trying to parse each field individually.